File: java/Tutorials/Media/Java/read-video-from-device/src/main/java/com/neurotec/tutorials/ReadVideoFromDevice.java
package com.neurotec.tutorials;

import java.io.IOException;
import java.util.EnumSet;

import com.neurotec.images.NImage;
import com.neurotec.lang.NThrowable;
import com.neurotec.licensing.NLicense;
import com.neurotec.media.NAudioFormat;
import com.neurotec.media.NMediaFormat;
import com.neurotec.media.NMediaReader;
import com.neurotec.media.NMediaSource;
import com.neurotec.media.NMediaType;
import com.neurotec.media.NVideoFormat;
import com.neurotec.tutorials.util.LibraryManager;
import com.neurotec.tutorials.util.Utils;

public final class ReadVideoFromDevice {

	private static final String DESCRIPTION = "Demonstrates capturing frames from device.";
	private static final String NAME = "read-video-from-device";
	private static final String VERSION = "9.0.0.0";

	private static void usage() {
		System.out.println("usage:");
		System.out.format("\t%s [frameCount]", NAME);
		System.out.println();
		System.out.println("\tframeCount - number of frames to capture from each device to current directory");
		System.out.println();
	}

	public static void main(String[] args) {
		final String Components = "Media";

		LibraryManager.initLibraryPath();

		Utils.printTutorialHeader(DESCRIPTION, NAME, VERSION, args);

		if (args.length < 1) {
			usage();
			System.exit(-1);
		}

		try {
			if (!(NLicense.obtainComponents("/local", 5000, Components))) {
				System.err.println(String.format("Could not obtain licenses for components: %s", Components));
				System.exit(-1);
			}

			int frameCount = Integer.parseInt(args[0]);
			if (frameCount == 0) {
				System.err.println("no frames will be captured as frame count is not specified");
			}

			System.out.println("quering connected video devices ...");
			NMediaSource[] devices = NMediaSource.enumDevices(EnumSet.of(NMediaType.VIDEO));
			System.out.format("devices found: %s\n", devices.length);

			for (NMediaSource source : devices) {
				System.out.format("found device: %s\n", source.getDisplayName());
				readFrames(source, frameCount);
				System.out.println("done");
			}
			System.out.println("done");
		} catch (Throwable th) {
			th.printStackTrace();

			int errorCode = -1;

			if (th instanceof NThrowable) {
				errorCode = ((NThrowable)th).getCode();
			}

			System.exit(errorCode);
		} finally {
			try {
				NLicense.releaseComponents(Components);
			} catch (IOException e) {
				e.printStackTrace();
			}
		}
	}

	private static void dumpMediaFormat(NMediaFormat mediaFormat) {
		if (mediaFormat == null) throw new NullPointerException("mediaFormat");

		switch (mediaFormat.getMediaType()) {
		case VIDEO:
			NVideoFormat videoFormat = (NVideoFormat) mediaFormat;
			System.out.format("video format .. %sx%s @ %s/%s (interlace: %s, aspect ratio: %s/%s)\n", videoFormat.getWidth(), videoFormat.getHeight(), videoFormat.getFrameRate().numerator, videoFormat.getFrameRate().denominator, videoFormat.getInterlaceMode(), videoFormat.getPixelAspectRatio().numerator, videoFormat.getPixelAspectRatio().denominator);
			break;
		case AUDIO:
			NAudioFormat audioFormat = (NAudioFormat) mediaFormat;
			System.out.format("audio format .. channels: %s, samples/second: %s, bits/channel: %s\n", audioFormat.getChannelCount(), audioFormat.getSampleRate(), audioFormat.getBitsPerChannel());
			break;
		default:
			throw new IllegalArgumentException("unknown media type specified in format!");
		}
	}

	private static void readFrames(NMediaSource mediaSource, int frameCount) throws Exception {
		NMediaReader mediaReader = null;
		try {
			mediaReader = new NMediaReader(mediaSource, EnumSet.of(NMediaType.VIDEO), true);

			System.out.format("media length: %s\n", mediaReader.getLength());

			NMediaFormat[] mediaFormats = mediaSource.getFormats(NMediaType.VIDEO);
			if (mediaFormats == null) {
				System.out.println("formats are not yet available (should be availble after media reader is started");
			} else {
				System.out.format("format count: %s\n", mediaFormats.length);
				for (int i = 0; i < mediaFormats.length; i++) {
					System.out.printf("[%s] ", i);
					dumpMediaFormat(mediaFormats[i]);
				}
			}

			NMediaFormat currentMediaFormat = mediaSource.getCurrentFormat(NMediaType.VIDEO);
			if (currentMediaFormat != null) {
				System.out.println("current media format:");
				dumpMediaFormat(currentMediaFormat);

				if (mediaFormats != null) {
					System.out.println("set the last supported format (optional) ... ");
					mediaSource.setCurrentFormat(NMediaType.VIDEO, mediaFormats[mediaFormats.length - 1]);
				}
			} else {
				System.out.println("current media format is not yet available (will be availble after media reader start)");
			}

			System.out.print("starting capture ... ");
			mediaReader.start();
			System.out.println("capture started");

			try {
				currentMediaFormat = mediaSource.getCurrentFormat(NMediaType.VIDEO);
				if (currentMediaFormat == null) {
					throw new Exception("current media format is not set even after media reader start!");
				}
				System.out.println("capturing with format: ");
				dumpMediaFormat(currentMediaFormat);

				for (int i = 0; i < frameCount; i++) {
					NMediaReader.ReadResult result = mediaReader.readVideoSample();
					NImage image = result.getImage();
					if (result.getImage() == null) return; // end of stream

					String filename = String.format("%d.jpg", i);
					image.save(filename);

					System.out.format("[%s %s] %s\n", result.getTimeStamp(), result.getDuration(), filename);
				}
			} finally {
				mediaReader.stop();
			}
		} finally {
			if (mediaReader != null) mediaReader.dispose();
		}
	}
}

File: java/Tutorials/Media/C/ReadVideoFromDevice/ReadVideoFromDevice.c
#include <TutorialUtils.h>

#ifdef N_MAC_OSX_FRAMEWORKS
	#include <NCore/NCore.h>
	#include <NMedia/NMedia.h>
	#include <NLicensing/NLicensing.h>
#else
	#include <NCore.h>
	#include <NMedia.h>
	#include <NLicensing.h>
#endif

const NChar title[] = N_T("ReadVideoFromDevice");
const NChar description[] = N_T("Demonstrates capturing frames from device.");
const NChar version[] = N_T("9.0.0.0");
const NChar copyright[] = N_T("Copyright (C) 2011-2017 Neurotechnology");

int usage()
{
	printf(N_T("usage:\n"));
	printf(N_T("\t%s [frameCount]\n"), title);
	printf(N_T("\n"));
	printf(N_T("\tframeCount - number of frames to capture from each device to current directory\n"));
	printf(N_T("\n"));
	return 1;
}

static NResult ReadFrames(HNMediaSource hSource, NInt frameCount, NInt deviceNumber);

int main(int argc, NChar **argv)
{
	NResult result = N_OK;
	NInt frameCount = 0;
	NInt i;
	const NChar * components = N_T("Media");
	NBool available = NFalse;
	HNMediaSource * arhDevices = NULL;
	NInt deviceCount = 0;
	HNMediaReader hReader = NULL;
	HNString hDisplayName = NULL;

	OnStart(title, description, version, copyright, argc, argv);

	if (argc < 2)
	{
		OnExit();
		return usage();
	}
		
	// check the license first
	result = NLicenseObtainComponents(N_T("/local"), N_T("5000"), components, &available);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("NLicenseObtainComponents() failed, result = %d\n"), result);
		goto FINALLY;
	}

	if (!available)
	{
		printf(N_T("Licenses for %s not available\n"), components);
		result = N_E_NOT_ACTIVATED;
		goto FINALLY;
	}
	
	frameCount = atoi(argv[1]);
	if (frameCount == 0)
	{
		printf(N_T("no frames will be captured as frame count is not specified"));
	}

	printf(N_T("quering connected video devices ...\n"));
	result = NMediaSourceEnumDevices(nmtVideo, 0, &arhDevices, &deviceCount);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to query connected video devices (result = %d)\n"), result);
		goto FINALLY;
	}
	printf(N_T("done\n"));

	printf(N_T("devices found: %d\n"), deviceCount);

	/** enumerate cameras */
	for (i = 0; i < deviceCount; ++i)
	{
		HNMediaSource hSource = arhDevices[i];

		/** get device settings */
		result = NMediaSourceGetDisplayNameN(hSource, &hDisplayName);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("failed to get device name (result = %d)!\n"), result);
			goto FINALLY;
		}
		{
			const NChar * szDisplayName;
			result = NStringGetBuffer(hDisplayName, NULL, &szDisplayName);
			if (NFailed(result))
			{
				result = PrintErrorMsgWithLastError(N_T("failed to get device name (result = %d)!\n"), result);
				goto FINALLY;
			}
			printf(N_T("found device: %s\n"), szDisplayName);
		}
		result = NStringSet(NULL, &hDisplayName);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("failed to clear display name (result = %d)!\n"), result);
			goto FINALLY;
		}

		result = NMediaReaderCreate(hSource, nmtVideo, NTrue, 0, &hReader);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("failed to create media reader from media source (result = %d)!\n"), result);
			goto FINALLY;
		}

		ReadFrames(hReader, frameCount, i);
		
		result = NObjectSet(NULL, &hReader);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("failed to clear reader (result = %d)!\n"), result);
			goto FINALLY;
		}
		printf(N_T("done\n"));
	}
	printf(N_T("done\n"));

	result = N_OK;
FINALLY:
	{
		NResult result2 = NObjectSet(NULL, &hReader);
		if (NFailed(result2)) PrintErrorMsg(N_T("NObjectSet() failed, result = %d\n"), result2);
		result2 = NObjectUnrefArray(arhDevices, deviceCount);
		if (NFailed(result2)) PrintErrorMsg(N_T("NObjectUnrefArray() failed, result = %d\n"), result2);
		result2 = NStringSet(NULL, &hDisplayName);
		if (NFailed(result2)) PrintErrorMsg(N_T("NStringSet() failed, result = %d\n"), result2);

		result2 = NLicenseReleaseComponents(components);
		if (NFailed(result2)) PrintErrorMsg(N_T("NLicenseReleaseComponents() failed, result = %d\n"), result2);
	}

	OnExit();

	return result;
}

static void DumpMediaFormat(HNMediaFormat hFormat)
{
	NMediaType formatMediaType;
	NResult result;

	if (hFormat == NULL) return;

	result = NMediaFormatGetMediaType(hFormat, &formatMediaType);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to get media type for specified format (result = %d)!\n"), result);
		return;
	}

	switch (formatMediaType)
	{
	case nmtVideo:
		{
			NUInt width, height;
			NURational frameRate;
			NVideoInterlaceMode interlaceMode;
			NURational aspectRatio;

			result = NVideoFormatGetWidth(hFormat, &width);
			if (NFailed(result))
			{
				result = PrintErrorMsgWithLastError(N_T("failed to get width for specified format (result = %d)!\n"), result);
				return;
			}
			result = NVideoFormatGetHeight(hFormat, &height);
			if (NFailed(result))
			{
				result = PrintErrorMsgWithLastError(N_T("failed to get height for specified format (result = %d)!\n"), result);
				return;
			}
			result = NVideoFormatGetFrameRate(hFormat, &frameRate);
			if (NFailed(result))
			{
				result = PrintErrorMsgWithLastError(N_T("failed to get frame rate for specified format (result = %d)!\n"), result);
				return;
			}
			result = NVideoFormatGetInterlaceMode(hFormat, &interlaceMode);
			if (NFailed(result))
			{
				result = PrintErrorMsgWithLastError(N_T("failed to get interlace mode for specified format (result = %d)!\n"), result);
				return;
			}
			result = NVideoFormatGetPixelAspectRatio(hFormat, &aspectRatio);
			if (NFailed(result))
			{
				result = PrintErrorMsgWithLastError(N_T("failed to get aspect ratio for specified format (result = %d)!\n"), result);
				return;
			}
			printf(N_T("video format .. %dx%d @ %d/%d (interlace: %d, aspect ratio: %d/%d)\n"), width, height,
				frameRate.Numerator, frameRate.Denominator, interlaceMode, aspectRatio.Numerator, aspectRatio.Denominator);
		}
		break;
	case nmtAudio:
		{
			NInt channelCount, sampleRate, bitsPerChannel;

			result = NAudioFormatGetChannelCount(hFormat, &channelCount);
			if (NFailed(result))
			{
				result = PrintErrorMsgWithLastError(N_T("failed to get channel count for specified format (result = %d)!\n"), result);
				return;
			}
			result = NAudioFormatGetSampleRate(hFormat, &sampleRate);
			if (NFailed(result))
			{
				result = PrintErrorMsgWithLastError(N_T("failed to get sample rate for specified format (result = %d)!\n"), result);
				return;
			}
			result = NAudioFormatGetBitsPerChannel(hFormat, &bitsPerChannel);
			if (NFailed(result))
			{
				result = PrintErrorMsgWithLastError(N_T("failed to get bits per sample for specified format (result = %d)!\n"), result);
				return;
			}
			printf(N_T("audio format .. channels: %d, samples/second: %d, bits/channel: %d\n"),
					channelCount, sampleRate, bitsPerChannel);
		}
		break;
	default:
		printf(N_T("unknown media type specified in format!\n"));
		return;
	}
}

static NResult ReadFrames(HNMediaReader hReader, NInt frameCount, NInt deviceNumber)
{
	HNImage hImage = NULL;
	NResult result;
	HNMediaFormat * arFormats = NULL;
	HNMediaFormat hCurrentFormat = NULL;
	NLong length;
	NInt formatCount = 0, j;
	HNMediaSource hSource = NULL;
	NBool finishCapture = NFalse;

	result = NMediaReaderGetSource(hReader, &hSource);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to get media source from media reader (result = %d)!\n"), result);
		goto FINALLY;
	}

	result = NMediaReaderGetLength(hReader, &length);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to get media length from media reader (result = %d)!\n"), result);
		goto FINALLY;
	}
	printf(N_T("media length: %lld\n"), length);

	result = NMediaSourceGetFormats(hSource, nmtVideo, &arFormats, &formatCount);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to get media formats from media source (result = %d)!\n"), result);
		goto FINALLY;
	}
	printf(N_T("format count: %d\n"), formatCount);
	for (j = 0; j < formatCount; j++)
	{
		printf(N_T("[%d] "), j);
		DumpMediaFormat(arFormats[j]);
	}

	result = NMediaSourceGetCurrentFormat(hSource, nmtVideo, &hCurrentFormat);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to get media formats from media source (result = %d)!\n"), result);
		goto FINALLY;
	}
	if (hCurrentFormat)
	{
		printf(N_T("current media format:\n"));
		DumpMediaFormat(hCurrentFormat);

		printf(N_T("set the last supported format (optional) ... "));
		result = NMediaSourceSetCurrentFormat(hSource, nmtVideo, arFormats[formatCount - 1]);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("failed to get media formats from media source (result = %d)!\n"), result);
			goto FINALLY;
		}
		printf(N_T("format setted\n"));
	}
	else
	{
		printf(N_T("current media format is not yet available (will be availble after media reader start)\n"));
	}

	result = NObjectSet(NULL, &hCurrentFormat);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to clear format (result = %d)!\n"), result);
		goto FINALLY;
	}

	printf(N_T("starting capture ... "));
	result = NMediaReaderStart(hReader);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to start media reader (result = %d)!\n"), result);
		goto FINALLY;
	}
	finishCapture = NTrue;
	printf(N_T("capture started\n"));

	result = NMediaSourceGetCurrentFormat(hSource, nmtVideo, &hCurrentFormat);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("failed to get current format (result = %d)!\n"), result);
		goto FINALLY;
	}
	if (!hCurrentFormat)
	{
		printf(N_T("current media format is not set even after media reader start!\n"));
		goto FINALLY;
	}
	DumpMediaFormat(hCurrentFormat);

	for (j = 0; j < frameCount; j++)
	{
		NLong timespan = -1, duration = 0;
		NChar szFilename[1024];
		
		result = NMediaReaderReadVideoSample(hReader, &timespan, &duration, &hImage);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("failed to read video sample from media reader (result = %d)!\n"), result);
			goto FINALLY;
		}
		if (!hImage)
			break;

		printf(N_T("[%llu %llu]\n"), timespan, duration);
		sprintf(szFilename, N_T("%d_%d.bmp"), deviceNumber, j);
		result = NImageSaveToFileEx(hImage, szFilename, NULL, NULL, 0);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("failed to save image (result = %d)!\n"), result);
			goto FINALLY;
		}
		result = NObjectSet(NULL, &hImage);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("failed to clear frame image (result = %d)!\n"), result);
			goto FINALLY;
		}

		fflush(stdout);
	}

	result = N_OK;
FINALLY:
	if (finishCapture)
	{
		result = NMediaReaderStop(hReader);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("failed to stop media reader (result = %d)!\n"), result);
		}
	}
	{
		NResult result2;
		result2 = NObjectSet(NULL, &hImage);
		if (NFailed(result2)) PrintErrorMsg(N_T("failed to clear sound buffer (result = %d)!\n"), result2);
		result2 = NObjectSet(NULL, &hCurrentFormat);
		if (NFailed(result2)) PrintErrorMsg(N_T("failed to clear current format (result = %d)!\n"), result2);
		result2 = NObjectUnrefArray(arFormats, formatCount);
		if (NFailed(result2)) PrintErrorMsg(N_T("failed to clear formats (result = %d)!\n"), result2);
		result2 = NObjectSet(NULL, &hSource);
		if (NFailed(result2)) PrintErrorMsg(N_T("failed to clear source (result = %d)!\n"), result2);
	}
	return result;
}
